kernel: Add an "identifier" to AppId

bradjc opened the pull request:
### Pull Request Overview

Currently a capsule can store an `AppId` to refer to a process, and there is not really a mechanism for ensuring that that `AppId` is still valid. Currently in Tock we are OK with this since we only create processes once, they don't restart, and we don't move them around in the processes array. However, as process management becomes more sophisticated, the idea that an AppId is always valid will no longer be true. Since `AppId` only stores the index of the relevant process in that array, if a process restarts, or if a process is removed and then a new one is added in the same slot, then the stored `AppId` will appear valid, even though it actually refers to a different process.

To enable verifying that `AppId`s are still valid, this PR adds a new field (`identifier`) to the `AppId` object. This allows `AppId`s to be used as before (i.e. they are still copyable), but now operations using `AppId` can verify that the identifier in the AppId matches the identifier assigned to the process at a particular index.

AppId retains the `index` field which does correspond to the index in the processes array (at least while the AppId is valid). This allows us to retain direct lookups, as well as simplifies the IPC implementation.

#### Changes to AppId API

The existing `.idx()` function has been removed, and `.index()` and `.id()` have been added. `.index()` checks that the AppId is still valid before returning the index in the array of the corresponding app. `.id()` just returns the identifier, which can be passed to userspace or over a UART channel, for example.

These changes resulted in a couple changes to capsules which need `usize` identifiers for apps, and now must call `.id()`.

#### Changes to kernel APIs

The major change in the kernel (beyond AppId) is that functions which used to take an index now just take an `AppId`. Mostly this is a minor change, except for `kernel.process_map_or()`. This function runs a closure on a specific app if it is exists, and uses the index and identifier to verify the app is still valid.

Otherwise, IPC had to be updated quite a bit to handle the changes. I also took the opportunity to update how grant iters are implemented since this PR reduces the reliance on an app's identifier being its index in the process array.

#### How does this fix the capsule problem?

A capsule can still store an AppId, and that AppId can still become invalid. However, the main use of an AppId between a capsule and the kernel is to call `grant.enter(appid)`. Without this PR, if appid refers to an old process, and a new process has started in the same array index, `grant.enter(appid)` would succeed (after initializing a new grant region for the new process). Now, when `grant.enter()` calls `.kernel.process_map_or(appid)`, an error will be returned which will bubble back to the capsule.

#### What does this PR not do?

This PR just improves the mechanism for detecting app changes, but does not change the policy on how app identifiers are assigned. Identifiers remain equivalent to their index in the array.

I iterated on the design in this PR several times, and the changes became substantial without also changing how app identifiers are assigned.

This is part of #1082.

### Testing Strategy

This pull request was tested by running the IPC apps. More testing is needed.


### TODO or Help Wanted

n/a


### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

hudson-ayers commented on kernel/src/sched.rs:
```diff
@@ -70,17 +70,37 @@ impl Kernel {
-    /// will be returned. Otherwise the closure will executed and passed a
-    /// reference to the process.
-    crate fn process_map_or<F, R>(&self, default: R, process_index: usize, closure: F) -> R
+    /// not exist (i.e. it is `None` in the `processes` array, or has stopped,
```
Is it true that if a process has been stopped this will return `default`? From looking through it seems that a stopped process still has a valid and unchanged AppId, and this function does not seem to call anything that checks if a process has been stopped.

bradjc replied:
Ah ok good catch! This raises a case I hadn't really considered...what should happen if an app is in the `StoppedFaulted` state (or perhaps an `Ended` state if one were to exist)? The process is still there, and its the same process so its identifier probably shouldn't change, and certain capsules may want to be able to see that it exists (like process console). However, other capsules should be able to know it is no longer active (and will never be again). Perhaps `process_map_or()` should not change, but users of `process_map_or()` may need to check that the process is still active. In particular grant.enter() shouldn't succeed on an ended process.

In any case I will update the comment.

bradjc replied:
I updated the comment.

bradjc commented:
Added a commit to address the issue of what happens with a "stopped" process, in particular a process that cannot be executed again. I call this an "inactive" process, and this can happen if the process crashes and the kernel does not restart it, or (hypothetically) if the process exits itself and the kernel does not free its `Process` struct.

The issue in this case is to ensure that capsules, or anything that holds an `AppId` pointing to that app, can learn that the app is inactive and will not run again and should not hold any state for it. Since the process is still valid in the processes array, calling `kernel.process_map_or()` will find a valid match, so that will not signal the caller that the process is inactive. I thought about requiring callers to check if the process is active, but that seems easy to forget to todo.

Instead, I went through all of the functions in the `ProcessType` trait and made it clear that any functions that attempt to modify a process in any way will fail if the process is inactive. That allows something to still query the process (say to access its debug information or for process console to show that the app exists and is inactive), but any attempt by an `AppId` holder to modify the process will cause an error to propagate.

The canonical example of this is when accessing a grant. Capsules access a specific grant by calling `grant.enter(appid)`. If the `appid` is completely invalid then when the grant module tries to find the process an error will be returned (as described above). However, if the appid is valid but the process is inactive, then the grant module will find the process but when it tries to manipulate the grant region an error will occur and that error will propagate back to the capsule.

ppannuto commented on kernel/src/callback.rs:
```diff
@@ -9,44 +9,122 @@ use crate::process;
+    /// originally referred to, even if that app no longer exists. For example,
+    /// the app may have restarted, or may have been ended or removed by the
+    /// kernel. Therefore, calling `id()` is _not_ a valid way to check
+    /// that an application still exists.
```
Thinking about global uniqueness of identifiers -- since we're proposing this as a mechanism that labels a process possibly off-core (here tockloader is given as example, but one would also imagine using as an RPC identifier), I think part of the future design discussion for the actual identifiers should include whether we have an entropy source of some kind in the kernel to try to ensure that identifiers are unique even across reboots

gendx replied:
> [...] whether we have an entropy source of some kind in the kernel to try to ensure that identifiers are unique even across reboots

Note that this would require more bits than `usize` from a cryptographic perspective, if you really want to avoid collision in all realistic scenarios (even considering that embedded CPUs are "slow" enough that one can only generate so many identifiers per second). A collision within a set of random 32-bit identifiers requires [only around 2^16 identifiers](https://en.wikipedia.org/wiki/Birthday_problem).

I don't have much context on the use cases, so I guess it depends whether uniqueness is a *nice to have* property for e.g. debugging, or if a collision is to be avoided at all costs.

bradjc commented:
>     1. Whatever solution we choose for assigning the "identifier" field, we will probably need to be careful to avoid reuse of identifiers.

Yes. My current simple implementation is just to keep counting up, and hope we don't hit 2^32. But we can discuss this more on the next PR.

>     2. It seems like this could potentially be implemented without storing the `index` field at all. This would require restructuring IPC to store tuples of (callbacks/identifiers) and (AppSlices/identifiers), and replacing the few places where we directly index into the process array with searches for the identifier. Given that no board currently supports more than 20 processes, this is not a huge overhead, though I realize this could be a limiting design decision if that were to change. I think the main overhead is that you would have to iterate through the list on every call to `.enter()`. The upside is that the resulting design is less confusing and entirely decouples processes from how the kernel stores them.

Yes, definitely, and my original implementation did just that in two different forms. It definitely complicates IPC, and I tried to store tuples of <identifier, callback/slice>. While conceptually it is pretty straightforward, the rust code gets much more complicated because there is a lot more checking required, with more `map()`s, etc. It eventually became too much of a burden to actually implement in Rust. My second version stored an array of identifiers, and used the index in the array of the identifier as the lookup to the arrays for callbacks and app slices. This was much easier to actually implement, and worked, but was a little clumsy since there are these extra arrays and a scan of those arrays. Plus, `process_map_or()` then had to be a scan, and that could potentially lead to nested scans if implementors aren't careful. Complicating IPC and forcing an array scan on every process lookup seemed like a rather large cost just to eliminate the index that we already have in AppId, hence I settled on the design you see in this PR.

bradjc commented on kernel/src/callback.rs:
```diff
@@ -9,44 +9,122 @@ use crate::process;
 impl PartialEq for AppId {
     fn eq(&self, other: &AppId) -> bool {
-        self.idx == other.idx
+        self.index == other.index && self.identifier == other.identifier
```
I think in the current implementation it is probably more correct to compare both fields. The reason is that using an appid with a valid identifier but the wrong index will not work, but using one with both fields correct will. It would be a little strange to have `AppIdA == AppIdB`, but then using `AppIdA` fails while `AppIdB` works.

At one point I did implement the logic for an AppId to automatically update its index, but since processes cannot currently move that seemed like unnecessary complexity right now.

ppannuto replied:
Hm, I'm not sure I'm convinced. The opposite argument here is that the `.id()` method is provided to
> retrieve a simple identifier for the process that can be communicated over a UART bus or syscall interface. This call is guaranteed to return a suitable identifier for the `AppId`

So now we have a situation where it's possible that `AppIdA.id() == AppIdB.id()` but `AppIdA != AppIdB`.

I feel like the identifier _is_ the identifier of an app, and the index is just an internal optimization that saves an otherwise expensive lookup.

What situation would arise where the identifiers match but the indexes don't?

bradjc replied:
> What situation would arise where the identifiers match but the indexes don't?

The kernel moves the application to a different index in the processes array. But, there isn't any functionality to do this, so it's a case that doesn't matter.

The fact that processes cannot move slots in the array (since there is currently no code that will do that) can support either implementation of `eq`. However, since just using `identifier` would be required if there ever was a reason to move processes around in the array, that implementation makes more sense. I will update.

hudson-ayers commented:
I came across a pretty obscure piece of code affected by this change while looking into modifying the ProcessConsole. Currently, the ProcessConsole lists process in the form:
```
 PID    Name                Quanta  Syscalls  Dropped Callbacks  Restarts    State
  00    hello3                 298        60                  0         0  Running
  01    hello2                 280        54                  0         0  Running
  02    hello_loop             262        99                  0         0  Running
```
where the PID is not actually the index of the process in the process array, but the index of the iterator through the processes array (as output by `.iter().enumerate()` in `process_each_capability()` in `grant.rs`). This means that if there are empty slots in the process array prior to an process, the PID assigned by the process console will match neither the identifier of the process nor its index in the process array. 

This is unchanged by this PR, but probably it would make more sense to have the ProcessConsole print the actual identifier of the process now that we have them.

bradjc commented:
I've tested this with various apps on Hail, including the tutorials/05 IPC apps, the Hail test app, and other collections of apps. I also tried these five:

```txt
 PID    Name                Quanta  Syscalls  Dropped Callbacks  Restarts    State
  00	hello_loop               0       359                  0         0  Yielded
  01	button_print             0         7                  0         0  Yielded
  02	adc_continuous           0       675                  0         0  Yielded
  03	rot13_client             0       829                  0         0  Yielded
  04	org.tockos.ex.rot13      0       131                  0         0  Yielded
```
and all 5 seem to work.

ppannuto commented on kernel/src/process.rs:
```diff
@@ -769,6 +829,11 @@ impl<C: Chip> ProcessType for Process<'a, C> {
     }
 
     fn setup_mpu(&self) {
+        // Do not modify an inactive process.
```
I'm curious about the proliferation of these:
 - Are all the functions with these checks actually getting called on stopped processes?
 - Should that be a bug rather than a silent ignore?
 - How do we make sure we don't miss these in future process management stuff?

bradjc replied:
>     * Are all the functions with these checks actually getting called on stopped processes?

Well the one that initiated this change is, notably if a capsule holds a copy of an appid for a stopped process and then uses it to enter its grant region. The others are not as most are called as a result of the process being scheduled, which it won't since it is stopped.

>     * Should that be a bug rather than a silent ignore?

In the case of entering a grant region it returns an error to the capsule.

For the others, maybe? It adds complexity in the common case. Maybe the better approach (at least for now) is to make the documentation clear that calling the functions without return codes is only valid when the process is active.



>     * How do we make sure we don't miss these in future process management stuff?

I'm not sure. Generally, functions with return values need to handle this, as those return values likely are passed (eventually) outside of the kernel crate. Functions without return values tend to be only called inside the kernel.

This is also complicated a little by code like the process console which does obtain references to `ProcessType` objects. But that is really a separate issue.

bradjc commented:
I rebased, and updated the AppId equality check, and undid some of the changes to the ProcessType trait and instead added a little more documentation for some methods.

bradjc commented:
Seems reasonable to add a last-call tag for this.

bradjc commented:
bors r+
